### PR TITLE
MsQueue to channel

### PR DESCRIPTION
**Pull request by petr-tik**
Implements own Queue using crossbeam-channel. Relies on blocking behaviour of the pop() method

Related to #485 and #491

### CODE REVIEW

File: src/core/pool.rs [hunk 1]
```diff
@@ -146,4 +177,45 @@ mod tests {
...
+        let q = Queue::new();
+        let elem = 5;
+        q.push(elem);
+        let res = q.pop();
+        assert_eq!(res.unwrap(), elem);
+    }
+
+    #[test]
+    fn test_queue_pushes_and_pops_order() {
+        let elements = vec![1, 2, 3, 4];
```
Comment (fulmicoton) [inline src/core/pool.rs]: There is no need to test for order actually.

File: src/core/pool.rs [hunk 2]
```diff
@@ -146,4 +177,45 @@ mod tests {
...
+    //         let elements_for_pool = vec![1, 2];
+    //         // Clone vector, so we can calculate its length later
+    //         pool.publish_new_generation(elements_for_pool.clone());
+
+    //         let mut threads = vec![];
+    //         let sleep_dur = time::Duration::from_millis(10);
+    //         // spawn one more thread than there are elements in the pool
+    //         for _thread_idx in 0..elements_for_pool.len() + 1 {
+    //             threads.push(thread::spawn(move || {
+    //                 pool.acquire();
```
Comment (fulmicoton) [inline src/core/pool.rs]: The point of this test is to see if the pool behaves correctly when the pool is exhausted.

Right now, your leased item is dropped  right away (before the call to thread::sleep()).
Please fix as follows:

`let _leased_searcher = pool.acquired();`

**Review by fulmicoton — CHANGES_REQUESTED** [review]

**Comment by petr-tik** [comment]
What about RecvError and dealing with Result from .send()?

**Comment by fulmicoton** [comment]
> What about RecvError and dealing with Result from .send()?

According to doc, they only exist to express that the respective other side of the pipe was dropped.
By design it cannot happen in that pool, so it's ok for you to `panic` on errors.

Just use `.expect()` in both cases.

File: src/core/pool.rs [hunk 3]
```diff
@@ -9,21 +10,50 @@ pub struct GenerationItem<T> {
     item: T,
 }
 
+/// Queue to replace the Michael-Scott Queue, now deprecated in crossbeam
```
Comment (fulmicoton) [inline src/core/pool.rs]: no need for this comment. People don't need to know the history of this queue.

File: src/core/pool.rs [hunk 4]
```diff
@@ -9,21 +10,50 @@ pub struct GenerationItem<T> {
...
+        Queue {
+            sender: s,
+            receiver: r,
+        }
+    }
+
+    /// Sender trait returns a Result type, which is ignored.
+    /// The Result is not handled at the moment
+    fn push(&self, elem: T) {
+        self.sender.send(elem);
```
Comment (fulmicoton) [inline src/core/pool.rs]: This generates a warning. Make sure you remove all warning at compilation.

**Comment by petr-tik** [comment]
Rebased onto master and _force_-pushed (sorry). 

thanks for your comments, I addressed them to the best of my ability.

**Comment by fulmicoton** [comment]
Thanks!

**Pull request merged**